components/scream/src/physics/p3/p3_functions_ice_collection_impl.hpp

```diff
@@ -13,31 +13,32 @@ ::ice_cldliq_collection(const Spack& rho, const Spack& temp,
                         const Spack& rhofaci, const Spack& f1pr04,
                         const Spack& qitot_incld, const Spack& qc_incld,
                         const Spack& nitot_incld, const Spack& nc_incld,
-                        Spack& qccol, Spack& nccol, Spack& qcshd, Spack& ncshdc)
+                        Spack& qccol, Spack& nccol, Spack& qcshd, Spack& ncshdc,
+                        const Smask& context)
 {
-   constexpr Scalar qsmall = C::QSMALL;
-   constexpr Scalar tmelt = C::Tmelt;
-
-   // set up masks
-   const auto t_is_negative        = temp < tmelt;
-   const auto qitot_incld_gt_small = qitot_incld > qsmall;
-   const auto qc_incld_gt_small    = qc_incld > qsmall;
-   const auto both_gt_small        = qitot_incld_gt_small && qc_incld_gt_small;
-   const auto both_gt_small_pos_t  = both_gt_small && !t_is_negative;
-
-   constexpr auto eci = C::eci;
-   constexpr auto inv_dropmass = C::ONE/C::dropmass;
-
-   qccol.set(both_gt_small && t_is_negative,
-             rhofaci*f1pr04*qc_incld*eci*rho*nitot_incld);
-   nccol.set(both_gt_small, rhofaci*f1pr04*nc_incld*eci*rho*nitot_incld);
-
-   // for T > 273.15, assume cloud water is collected and shed as rain drops
-   // sink for cloud water mass and number, note qcshed is source for rain mass
-   qcshd.set(both_gt_small_pos_t, rhofaci*f1pr04*qc_incld*eci*rho*nitot_incld);
-   nccol.set(both_gt_small_pos_t, rhofaci*f1pr04*nc_incld*eci*rho*nitot_incld);
-   // source for rain number, assume 1 mm drops are shed
-   ncshdc.set(both_gt_small_pos_t, qcshd*inv_dropmass);
+  constexpr Scalar qsmall = C::QSMALL;
+  constexpr Scalar tmelt = C::Tmelt;
+
+  // set up masks
+  const auto t_is_negative        = temp < tmelt;
+  const auto qitot_incld_gt_small = qitot_incld > qsmall;
+  const auto qc_incld_gt_small    = qc_incld > qsmall;
+  const auto both_gt_small        = qitot_incld_gt_small && qc_incld_gt_small;
+  const auto both_gt_small_pos_t  = both_gt_small && !t_is_negative;
+
+  constexpr auto eci = C::eci;
+  constexpr auto inv_dropmass = C::ONE/C::dropmass;
+
+  qccol.set(both_gt_small && t_is_negative && context,
```

Comment:
Here and in the following, there are some mask op redundancy. E.g, maybe precompute the mask `both_gt_small && context` only once?

Reply:
Thanks, this is a nice cleanup and more consistent with the approach I tried to use in other functions.